### ISSUE TITLE
Some Active Sessions show as Worlds in the browser

### ISSUE THREAD

**Issue by shadowpanther** [Yellow-Dog-Man/Resonite-Issues#164]
### Describe the bug?

While looking at the Active Sessions tab of the World Browser, some shown sessions don't show user count and the hosting user, but show world creator instead. When I click to get the world info, the popup shows it's a session and shows the host and the users in session.

### To Reproduce

Look at the Active Worlds tab in World Browser.
Notice after some time the lack of user count indicator and the Host line changes to By line.

### Expected behavior

Active sessions show as sessions, not as worlds that have no sessions of them,

### Screenshots

![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/675104/c372449a-1b9d-4829-80e7-44c1e66ae700)
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/675104/0273b5b3-5476-4a37-9b4f-a896af7144ab)
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/675104/4010c518-e6ca-44a8-ad8d-05cb63763224)
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/675104/070da21f-8076-461d-ad2b-663db75b58b9)

### Resonite Version Number

2023.10.13.743

### What Platforms does this occur on?

Windows

### Log Files

[VIXEN - 2023.10.13.743 - 2023-10-14 17_35_20.log](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/12906632/VIXEN.-.2023.10.13.743.-.2023-10-14.17_35_20.log)

### Additional Context

~~My guess is the sessions shown this way might be the ones where more than one session of the same public world is open.~~
Checked again, and it seems even the unique session of one world is sometimes shown this way.

### Reporters

Shadow Panther

**Comment by shadowpanther** [comment]
There's also an opposite effect sometimes, when the session shows in the list, but the session info is empty:
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/675104/8581cdd7-63b3-4486-8889-4a13e5984de9)
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/675104/228e61a2-86ac-4d6b-9ffe-ebbd3f5e8212)
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/675104/389e68bc-476c-4694-aefe-d922d984e50d)

**Comment by Nytra** [comment]
The bug where the session shows in the list but upon clicking shows no session info is extremely annoying, because at this point the only way you have to join such a session is via the contacts list, but only if you are contacts with anyone in that session.

**Comment by Nytra** [comment]
I have a feeling about what causes the second problem in this issue: "session shows in the list, but the session info is empty"

I think it happens if you host a session of a world yourself. Somehow I think that session info gets cached. Then later if you try to click on a active session of that world in the browser it uses the previously cached session that you hosted which no longer exists. There seems to be a cloud record cache that assigns to `WorldItem._worldRecord` which is used by `WorldItem.SetTarget` upon clicking the `WorldThumbnailItem`. However this is just a hypothesis and it would need to be tested more.

**Comment by Nytra** [comment]
I have found that if you untick `MergeSessionsByWorldId` in the WorldListManager component it fixes the problem of "session shows in the list, but the session info is empty"

**Comment by akiRAM3** [comment]
This issue is still ongoing.
Like Nytra, I have to use contact list to join my friend's session in my daily basis.

And, when it comes to JP Tutorial World, where multiple sessions tends to occur, 
it is very inconvinient to noticing/tracking sessions built by new people who are not in my contact list due to this bug.

**Comment by e1ght3** [comment]
I would like to see this fixed as soon as possible, as it is a hindrance to new encounters when sessions do not display properly during normal operation.

**Comment by skywindkitsune** [comment]
This could be related to issue #739 sometimes when you login to resonite you do not register as online automatically causing the world browser to not display the session as an active session.

**Comment by shadowpanther** [comment]
@skywindkitsune the missing session in Active sessions happens even if the contact shows as online, and I can join that same session from the Contacts tab. It's only in the world browser that the problem happens.

**Comment by Frooxius** [comment]
I'll try to look into this soon.

**Comment by Frooxius** [comment]
I'm looking into this, but there's not a super good spot to start for this that I can see.

Are there any patterns to this behavior? E.g.:

- Does this happen only with sessions that have been running for a while (order of hours)? Or does it also affect relatively fresh sessions (less than half an hour)?
- Does it affect any session access level or just specific ones? (Public, Contacts, Contacts+...)
- Does this seem to affect all users equally? E.g. if this occurs for particular session, will multiple people be affected at the same time? Or would some users see the session without an issue, while others have an issue?
- If you restart the client, does the issue persist for the same session?

**Comment by iamgreaser** [comment]
The only pattern at the moment that I can see is that at least one of the two versions of the bug is pretty much guaranteed to happen. Right now I'm getting the variant where a world on the active sessions list looks like it has no sessions but if you click on it then there is a session active.

AND JUST LIKE THAT I managed to get the other variant to occur.
Look at the JP Tutorial World in the session browser and leave it on that for a minute or so. That one tends to have issues. The exact text you're looking for is `JP チュートリアルワールド Resonite` .

**Comment by Nytra** [comment]
@Frooxius I don't have concrete data, but just from my experience with this I think it affects all sessions that have corresponding worlds, it seems to happen regardless of uptime or session access level, and I think it's a local problem so if person A sees the session disappear then person B might still be able to see it.

I can replicate it just by opening Active Sessions and watching it for under a minute, you will see sessions which are worlds begin to show only the world, and it can be fixed by refreshing the world browser (so clicking to some other tabs in the world browser and then back to Active Sessions). Restarting the client will 'fix' it in the same way as refreshing the browser I think, but then the sessions will start disappearing again.

---

Just for additional info:

I managed to 'fix' this in a [mod](https://github.com/Nytra/ResoniteFixSessionBrowser/blob/a470179802d252eb7018358602ae02ee7b3f06b2/FixSessionBrowser/FixSessionBrowser.cs#L179) by adding some code to the end of `WorldItem.OnWorldIDSessionsChanged` which calls the `WorldItem.ForceUpdate` method.

```
private void OnWorldIdSessionsChanged(IReadOnlyList<SessionInfo> sessions)
{
	RunSynchronously(delegate
	{
		UpdateData(sessions);
	});
        
        // Next part is the code I added
        RunSynchronously(delegate
	{
		ForceUpdate();
	});
}
```

I think part of the problem is that the sessions list that is getting provided into the `UpdateData` method is not containing the right sessions here for some reason, but calling `ForceUpdate` after it which gets a whole new sessions list is a way around it.

**Comment by Frooxius** [comment]
Thanks for the additional info and context! I'll see if I can figure something out. I tried watching the sessions for a bit too, but it didn't occur. 

It almost sounds like some data is expiring in some way or something like that, but I'm not sure. If that was the case I don't think ForceUpdate would work, so it might be that the data is not including the correct list of things.

**Comment by Frooxius** [comment]
Thanks! I was about to post that! Yes, please keep a lookout for this happening and then nab the logs and the name of the world so I can match up that to the log.

This doesn't fix the issue, but will give me more data to work with.

**Comment by Nytra** [comment]
@Frooxius I think I have an idea of what is actually causing this. Forgive the screenshot of decompiled code:

![Screenshot 2024-01-29 231242](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/14206961/c2a0f1c6-530b-41e8-bf14-9b931bb29651)

This is in `SkyFrost.Base.SessionsManager.RunSessionUpdated` method.

You can see that the code calls `item2(list3)` where `item2` is `FrooxEngine.WorldItem.OnWorldIdSessionsChanged`

It then immediately calls `Pool.Return(ref list3);` which clears the sessions list. This means that the sessions list is empty so when the synchronous action in `OnWorldIdSessionsChanged` tries to read from it there will be no sessions in there.

**Comment by ProbablePrime** [comment]
I'm unable to locate your exact code in our source, this is probably decompiling mangling.

I found the area though and the Pool use there is temporary, and is used to setup event listeners, It is not the main session list...list. So I do not think it is related.

**Comment by Nytra** [comment]
Well I patched the DLL to remove that call to `Pool.Return` and the bug no longer happens, so I think it is related. This is however not the proper solution because it will result in memory leaks over time.

**Comment by XDelta** [comment]
Spotted this occur on 'Optimizone' which has several sessions of the world being hosted. Occurred while I was looking at the 'contacts sessions' page and had opened up to the large info screen on the sessions available of that world.

Relevant log lines below
```12:49:29 AM.853 (125 FPS)	SessionInfo. Id: S-U-Raid-Server-Legends:OptimizoneLaRette, Name: Optimizone: LaRette, Host: Optimizone, CorrespondingWorldId: U-Raidriar796:R-f97cabfa-16c5-4d60-a371-25c0a50dbf42, URLs: lnl-nat://3384700975cb4fe78d99ba18430fcfba/S-U-Raid-Server-Legends:OptimizoneLaRette, IsExpired: False
12:49:29 AM.853 (125 FPS)	SessionInfo. Id: S-U-Raid-Server-Legends:OptimizoneMelody, Name: Optimizone: Melody, Host: Optimizone, CorrespondingWorldId: U-Raidriar796:R-f97cabfa-16c5-4d60-a371-25c0a50dbf42, URLs: lnl-nat://f4145623eceb433d90dd08e947cc0fc9/S-U-Raid-Server-Legends:OptimizoneMelody, IsExpired: False
12:49:29 AM.853 (125 FPS)	SessionInfo. Id: S-U-Raid-Server-Legends:OptimizoneNorland, Name: Optimizone: Norland, Host: Optimizone, CorrespondingWorldId: U-Raidriar796:R-f97cabfa-16c5-4d60-a371-25c0a50dbf42, URLs: lnl-nat://f4145623eceb433d90dd08e947cc0fc9/S-U-Raid-Server-Legends:OptimizoneNorland, IsExpired: False
12:49:29 AM.853 (125 FPS)	SessionInfo. Id: S-U-Raid-Server-Legends:OptimizoneRusset, Name: Optimizone: Russet, Host: Optimizone, CorrespondingWorldId: U-Raidriar796:R-f97cabfa-16c5-4d60-a371-25c0a50dbf42, URLs: lnl-nat://f4145623eceb433d90dd08e947cc0fc9/S-U-Raid-Server-Legends:OptimizoneRusset, IsExpired: False
12:49:29 AM.853 (125 FPS)	SessionInfo. Id: S-U-Raid-Server-Legends:OptimizoneYukon, Name: Optimizone: Yukon, Host: Optimizone, CorrespondingWorldId: U-Raidriar796:R-f97cabfa-16c5-4d60-a371-25c0a50dbf42, URLs: lnl-nat://f4145623eceb433d90dd08e947cc0fc9/S-U-Raid-Server-Legends:OptimizoneYukon, IsExpired: False
12:49:29 AM.853 (125 FPS)	SessionInfo. Id: S-U-Raid-Server-Legends:OptimizoneJewel, Name: Optimizone: Jewel, Host: Optimizone, CorrespondingWorldId: U-Raidriar796:R-f97cabfa-16c5-4d60-a371-25c0a50dbf42, URLs: lnl-nat://5ade3b968ba74298b186f508ca90c192/S-U-Raid-Server-Legends:OptimizoneJewel, IsExpired: False
12:49:31 AM.122 (126 FPS)	Session updated, forcing status update
12:49:31 AM.424 (126 FPS)	Session updated, forcing status update
12:49:31 AM.824 (140 FPS)	Session updated, forcing status update
12:49:32 AM.024 (146 FPS)	Session updated, forcing status update
12:49:37 AM.533 (143 FPS)	Session updated, forcing status update
12:49:37 AM.834 (143 FPS)	Session updated, forcing status update
12:49:38 AM.336 (141 FPS)	Session updated, forcing status update
12:49:39 AM.238 (144 FPS)	Session updated, forcing status update
12:49:40 AM.740 (144 FPS)	Session updated, forcing status update
12:49:41 AM.240 (144 FPS)	Session updated, forcing status update
12:49:42 AM.341 (144 FPS)	Session updated, forcing status update
12:49:42 AM.934 (144 FPS)	Session count became zero for U-Raidriar796:R-f97cabfa-16c5-4d60-a371-25c0a50dbf42 Optimizone

  at System.Environment.get_StackTrace () [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at Elements.Core.UniLog.Log (System.String message, System.Boolean stackTrace) [0x00000] in <80229d1197cf49d988a8b637774ba61f>:0 
  at FrooxEngine.WorldItem.UpdateData (System.Collections.Generic.IReadOnlyList`1[T] sessions, System.Boolean notifyManager) [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.WorldItem+<>c__DisplayClass19_0.<OnWorldIdSessionsChanged>b__0 () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.World.RunSynchronousActions () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.World.NextRefreshStage () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.World.Refresh () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.WorldManager.UpdateStep () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.WorldManager.RunUpdateLoop () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.Engine.UpdateStep () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.Engine.RunUpdateLoop () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at UnityFrooxEngineRunner.FrooxEngineRunner.UpdateFrooxEngine () [0x00000] in <5e585788751a4bbabeeadefebaee83eb>:0 
  at UnityFrooxEngineRunner.FrooxEngineRunner.Update () [0x00000] in <5e585788751a4bbabeeadefebaee83eb>:0 
12:49:42 AM.934 (144 FPS)	Session count became zero for U-Raidriar796:R-f97cabfa-16c5-4d60-a371-25c0a50dbf42 Optimizone

  at System.Environment.get_StackTrace () [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at Elements.Core.UniLog.Log (System.String message, System.Boolean stackTrace) [0x00000] in <80229d1197cf49d988a8b637774ba61f>:0 
  at FrooxEngine.WorldItem.UpdateData (System.Collections.Generic.IReadOnlyList`1[T] sessions, System.Boolean notifyManager) [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.WorldItem+<>c__DisplayClass19_0.<OnWorldIdSessionsChanged>b__0 () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.World.RunSynchronousActions () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.World.NextRefreshStage () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.World.Refresh () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.WorldManager.UpdateStep () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.WorldManager.RunUpdateLoop () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.Engine.UpdateStep () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.Engine.RunUpdateLoop () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at UnityFrooxEngineRunner.FrooxEngineRunner.UpdateFrooxEngine () [0x00000] in <5e585788751a4bbabeeadefebaee83eb>:0 
  at UnityFrooxEngineRunner.FrooxEngineRunner.Update () [0x00000] in <5e585788751a4bbabeeadefebaee83eb>:0 ```
  
Additional thing I noticed is it seems to occur more on worlds with multiple sessions open of it, although it definitely still occurs on worlds with a single instance open. Seems like whatever happens on any of the sessions will then hide all of them.

**Comment by XDelta** [comment]
Another separate occurance (on a world with one one open session)
```
  1:03:17 AM.834 (140 FPS)	SessionInfo. Id: S-44b4a8e7-ffbd-4269-b124-4b204baf17ed, Name: NeoRoid HUB WIP, Host: K-Pere, CorrespondingWorldId: U-KPere:R-e6481372-7844-4b90-b0fc-0bea4b87f095, URLs: res-steam://76561198079034428/2/S-44b4a8e7-ffbd-4269-b124-4b204baf17ed, lnl-nat://5ade3b968ba74298b186f508ca90c192/S-44b4a8e7-ffbd-4269-b124-4b204baf17ed, IsExpired: False
  1:03:17 AM.945 (140 FPS)	Session updated, forcing status update
  1:04:59 AM.091 (130 FPS)	Session updated, forcing status update
  1:05:06 AM.024 (136 FPS)	Session count became zero for U-KPere:R-e6481372-7844-4b90-b0fc-0bea4b87f095 NeoRoid HUB WIP

  at System.Environment.get_StackTrace () [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at Elements.Core.UniLog.Log (System.String message, System.Boolean stackTrace) [0x00000] in <80229d1197cf49d988a8b637774ba61f>:0 
  at FrooxEngine.WorldItem.UpdateData (System.Collections.Generic.IReadOnlyList`1[T] sessions, System.Boolean notifyManager) [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.WorldItem+<>c__DisplayClass19_0.<OnWorldIdSessionsChanged>b__0 () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.World.RunSynchronousActions () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.World.NextRefreshStage () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.World.Refresh () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.WorldManager.UpdateStep () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.WorldManager.RunUpdateLoop () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.Engine.UpdateStep () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at FrooxEngine.Engine.RunUpdateLoop () [0x00000] in <41c49157bb9444ac92cd54b5749246e7>:0 
  at UnityFrooxEngineRunner.FrooxEngineRunner.UpdateFrooxEngine () [0x00000] in <5e585788751a4bbabeeadefebaee83eb>:0 
  at UnityFrooxEngineRunner.FrooxEngineRunner.Update () [0x00000] in <5e585788751a4bbabeeadefebaee83eb>:0 ```

**Comment by Frooxius** [comment]
Thanks for all the info! Yes, I think I know why this is happening now, the lists are not being used right away, they are being returned to the pool (which clears them right after call finishes).

We'll just eat the cost of a little extra GC pressure (it shouldn't be much overall) and just allocate them fresh, which should fix this.

**Comment by Frooxius** [comment]
Okay I think this should be fixed now in 2024.1.30.731!

Thanks for all the help everyone! If this still occurs, we'll reopen, but I'm going to close it for now.
